File: leetcode/95/P95.rb
# 95. Unique Binary Search Trees II
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {Integer} n
# @return {TreeNode[]}
def generate_trees(n)
	return [] if n == 0
	search 1, n
end

# @return {TreeNode[]}
def search left, right
	list = []
	if left > right
		list << nil
		return list
	end	
	for i in left..right
		listLeft = search left, i - 1
		listRight = search i + 1, right
		listLeft.each do |treeLeft|
			listRight.each do |treeRight|
				root = TreeNode.new i
				root.left = treeLeft
				root.right = treeRight
				list << root
			end
		end
	end
	list
end

File: leetcode/95/src/P95.java
/**
 * Created by qwordy on 2/22/16.
 * 95. Unique Binary Search Trees II
 */

import java.util.ArrayList;
import java.util.List;

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;
	TreeNode(int x) { val = x; }
}

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
	public List<TreeNode> generateTrees(int n) {
		if (n == 0) return new ArrayList<>();
		return search(1, n);
	}

	private List<TreeNode> search(int left, int right) {
		List<TreeNode> list = new ArrayList<>();
		if (left > right) {
			list.add(null);
			return list;
		}
		for (int i = left; i <= right; i++) {
			List<TreeNode> listLeft = search(left, i - 1);
			List<TreeNode> listRight = search(i + 1, right);
			for (TreeNode treeLeft : listLeft) {
				for (TreeNode treeRight : listRight) {
					TreeNode root = new TreeNode(i);
					root.left = treeLeft;
					root.right = treeRight;
					list.add(root);
				}
			}
		}
		return list;
	}
}

public class P95 {
}
